src/client/pydaos/pydaos_shim.c

```diff
@@ -48,32 +52,172 @@ __is_magic_valid(int input)
 	return 1;
 }
 
+/* Macro that parses out magic value and verifies it */
+#define RETURN_NULL_IF_BAD_MAGIC(args)					\
+do {									\
+	int magic;							\
+	if (!PyArg_ParseTuple(args, "i", &magic)) {			\
+		DEBUG_PRINT("Bad arguments passed to %s", __func__);	\
+		return NULL;						\
+	}								\
+									\
+	if (!__is_magic_valid(magic)) {					\
+		return NULL;						\
+	}								\
+} while (0)
+
+
+/* Parse arguments and magic number out*/
+#define RETURN_NULL_IF_FAILED_TO_PARSE(args, format, x...)		\
+do {									\
+	int magic;							\
+	if (!PyArg_ParseTuple(args, "i"format, &magic, x)) {		\
+		D_DEBUG(DB_ANY, "Bad args passed to %s", __func__);	\
+		return NULL;						\
+	}								\
+									\
+	if (!__is_magic_valid(magic)) {					\
+		return NULL;						\
+	}								\
+} while (0)
+
 /**
- * Implementations of shim functions
+ * Implementations of baseline shim functions
  */
 
 static PyObject *
-__shim_handle__test(PyObject *self, PyObject *args)
+__shim_handle__daos_init(PyObject *self, PyObject *args)
+{
+	int rc;
+
+	rc = daos_init();
+
+	return PyInt_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__err_to_str(PyObject *self, PyObject *args)
 {
-	long ret;
-	int magic;
+	const char	*str;
+	int		 val;
 
-	if (!PyArg_ParseTuple(args, "i", &magic)) {
-		D_ERROR("Bad arguments passed to %s\n", __func__);
-		return NULL;
+	/* Parse arguments */
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "i", &val);
+	/* Call C function */
+	str = d_errstr(val);
+	if (str == NULL) {
+		Py_INCREF(Py_None);
+		return Py_None;
 	}
-	if (!__is_magic_valid(magic)) {
-		return NULL;
+
+	return PyString_FromString(str);
+}
+
+/**
+ * Implementation of container functions
+ */
+
+static PyObject *
+cont_open(int ret, uuid_t puuid, uuid_t cuuid, int flags)
+{
+	PyObject	*return_list;
+	daos_handle_t	 poh;
+	daos_handle_t	 coh;
+	d_rank_list_t	*svcl = NULL;
+	int		 rc;
+
+	if (ret != DER_SUCCESS) {
+		rc = ret;
+		goto out;
 	}
 
-	/* Call C function */
-	ret = 1;
+	svcl = daos_rank_list_parse("0", ":");
+	if (svcl == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	/** Connect to pool. XXX svcl is currently hardcoded */
+	rc = daos_pool_connect(puuid, "daos_server", svcl, DAOS_PC_RW, &poh,
+			       NULL, NULL);
+	if (rc)
+		goto out;
+
+	/** Open container */
+	rc = daos_cont_open(poh, cuuid, DAOS_COO_RW, &coh, NULL, NULL);
+	if (rc)
+		daos_pool_disconnect(poh, NULL);
+out:
+	if (svcl)
```

Comment:
(style)  suspect code indent for conditional statements (8, 17)

Reply:
fixed